### ISSUE TITLE
Unstable operation using the motor instead of the bulb

### ISSUE THREAD

**Issue by DukesPL** [AJMansfield/TriacDimmer#23]
Hi,
I have a problem controlling an AC motor using this library. Generally, the engine changes speed smoothly and works correctly both at minimum and extreme setBrightness values. But sometimes (I don't know how to write it exactly) it seems to lose synchronization and starts "firing", as when the engine is burned out or the brushes are running out. Sometimes it can fire once every few minutes, and sometimes several times in a row. I hope you know what I mean.
I tried using different values in the configuration (pulse_lenght, min_trigger, etc.) but it did not solve my problems. The engine is definitely functional because I tried it on several others and the effect was repeated. Maybe there is something that can be changed in the code to prevent this effect from occurring?

**Comment by NicBout** [comment]
I don't think a code change would solve the problem imho. The problem is more likely to be caused by voltage spikes from the coils in your AC motor. I think you might be able to fix your problem by adding a snubber (Resistor + capacitor) on the motor side of the circuit. Values depends on motor ratings, google should return more scientific explanations and links to a few calculators. I do this when using TRIACS with AC motors to prevent misfiring issues like the one you are describing, and avoid damage to the TRIACS...

**Comment by DukesPL** [comment]
I have already added a snubber and unfortunately it doesn't helped.

**Comment by AJMansfield** [comment]
It does seem likely that this is an electrical rather than software issue -- but it's an issue that seems to come up often enough here that it's probably worth figuring out finally so I can add a help section about it.

Do you have an oscilliscope that you can try to capture one of these events with? If I can get a look at the actual waveform here that'll dramatically accelerate the troubleshooting process (and I don't need some fancy data capture, a picture of the 'scope screen taken with your smartphone would be plenty).

Alternatively, can you provide a schematic of your circuit and part numbers/ratings for the motor, triac, and snubber network? If nothing else, I may try to replicate an equivalent setup in my lab.

**Comment by AJMansfield** [comment]
I should note: both the Krida and Robotdyn dimmer boards linked in the readme have a built-in 39 ohm / 10 nF snubber. Are you using one of those boards or is this your own triac circuit?

Depending on what sort of mains voltage you have and the current rating of the motor you might need different or additional snubbers, but that's something I'd need specific numbers for the triac/motor/etc in order to calculate.

**Comment by DukesPL** [comment]
https://github.com/AJMansfield/TriacDimmer/assets/37274843/e53b043c-b7d2-4d23-9e58-863feff9d3c3
![firing](https://github.com/AJMansfield/TriacDimmer/assets/37274843/17f7393b-1e04-4698-961c-f25e022f5cc1)

**Comment by DukesPL** [comment]
![image](https://github.com/AJMansfield/TriacDimmer/assets/37274843/d568092b-efa7-45eb-8bda-7d35fe577c46)

**Comment by DukesPL** [comment]
Part numbers are exactly as on the schematic

**Comment by AJMansfield** [comment]
Which signal are you recording with this?

If possible, could you capture it with the un-switched mains AC input, the switched output, and both the zero crossing and triac output pins? If you only have two channels, the part of the schematic I'm most suspicious of is the zero-crossing -- this seems to be a synchronization issue, so we want to make sure the sync signal itself isn't having trouble.

(Also, pro tip: most scopes have a "mains AC sync" triggering mode -- I suspect that using that will remove the jitter from that recording.)

**Comment by DukesPL** [comment]
> Which signal are you recording with this?
> 
> If possible, could you capture it with the un-switched mains AC input, the switched output, and both the zero crossing and triac output pins? If you only have two channels, the part of the schematic I'm most suspicious of is the zero-crossing -- this seems to be a synchronization issue, so we want to make sure the sync signal itself isn't having trouble.
> 
> (Also, pro tip: most scopes have a "mains AC sync" triggering mode -- I suspect that using that will remove the jitter from that recording.)

I have captured the signal at power output from pcb. To do that measure, I have used a transformer (230V/12V) to isolate oscilloscope from the main AC line.

I have 4ch oscilloscope, so today I will make some measures according to your instructions.

**Comment by DukesPL** [comment]
> If possible, could you capture it with the un-switched mains AC input, the switched output, and both the zero crossing and triac output pins?

Do you mean to connect scope just like on schematic bellow?
![image](https://github.com/AJMansfield/TriacDimmer/assets/37274843/37947f7f-9aa5-4bed-b42a-2b4105254b8b)

**Comment by AJMansfield** [comment]
> > If possible, could you capture it with the un-switched mains AC input, the switched output, and both the zero crossing and triac output pins?
> 
> So should I connect the oscilloscope as in the schematic below?

Apologies for the delay, yes those would be the points I'd recommend probing.

**Comment by AJMansfield** [comment]
Also, can you share the code you are using to set up and control the dimmer library?

**Comment by AJMansfield** [comment]
Anyway, I've written up a bit of an explanation here of the theory of operation of my library, in case any of this helps:

I'd highly recommend reading through the [Atmega328p datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-7810-Automotive-Microcontrollers-ATmega328P_Datasheet.pdf), specifically chapter 15: 16-bit Timer/Counter1 with PWM. That's the primary document that describes the operation of the PWM peripheral this library uses.

---

The library uses Timer/Counter 1 in free-running aka "Normal" mode, where the counter is allowed to count up continuously and overflow from 0xFFFF back to 0x0000. It's set to run from the system clock, with a divide-by-8 prescaler that ensures that the counter's 16 bits are enough to represent the period between zero-crossings and avoid aliasing between overflows.

With the most common 328p Arduino boards with external 16 MHz crystals, (e.g. an Uno, Micro, Nano, Leonardo), each count represents 0.5us, and the timer overflows back to 0 approximately every 33ms. On other boards that use the 328p's own internal 8 MHz oscillator, each count represents a nominal 1us, for an overflow about every 65ms -- but operating under conditions that put it closer to the edges of the ±14% tolerance, each count could represent anything from 0.86us to 1.14us, and the overflows could be anywhere from 57ms to 76ms.

For comparison, on 50Hz power there are 100 zero-crossings per second, so the time between those events is 10ms; whereas on 60Hz power there are 120 so the time between those events is 8.3ms. Either way, the time between events is enough smaller than the overflow period to avoid aliasing.

The core functionality of this library revolves around the interplay between the counter's Input Capture unit and the Output Compare units when used in free-running normal mode.

The Input Capture unit is a piece of hardware that, when triggered by an external pin change, instantly copies the current value of the counter to its `ICR1` register.
The Output Compare units have a similar function, but in reverse; they continuously wait and check for the counter to reach the value programmed into their `OCR1` register (`OCR1A` for unit A, `OCR1B1` for unit B), and then instantly change the value of the associated pin.

The value captured by the input capture unit acts as something of a timestamp; it can be compared to work out the time between events by subtracting a previous timestamp, or used to calculate a value to set the output compare register to to generate an output after a precise duration.

(Note that the "units" for these timestamps and durations are not necessarily exact 1us or 0.5us steps, they're just "counts". Thermal drift could even mean that the length of these counts could drift slightly over time -- but as long as it's slow enough the library should be able to automatically compensate as it happens.)

Both of those concepts -- comparing capture timestamps and computing output timestamps -- are used by this library. Probably the easiest-to-follow examples of this are in the [`TIMER1_CAPT` interrupt service routine](https://github.com/AJMansfield/TriacDimmer/blob/master/src/TriacDimmer.cpp#L147-L173).

The comparison between captured timestamps happens on [lines 163-164](https://github.com/AJMansfield/TriacDimmer/blob/master/src/TriacDimmer.cpp#L163-L164):
```
	TriacDimmer::detail::period = ICR1 - last_icr;
	last_icr = ICR1;
```

`TriacDimmer::detail::period` is a variable that the library uses to store the measured half-wave period, in timer-count units, and communicate that value from the ISR context to what I'll call the "userland" half of the library.

And calculating and setting the output timestamps happens a few lines earlier, [lines 152-153](https://github.com/AJMansfield/TriacDimmer/blob/master/src/TriacDimmer.cpp#L152-L153):
```
	OCR1A = ICR1 + TriacDimmer::detail::ch_A_up;
	OCR1B = ICR1 + TriacDimmer::detail::ch_B_up;
```

`TriacDimmer::detail::ch_A_up` and `ch_B_up` are variables that the library uses to store how long the pause between the zero crossing and start of trigger pulse should be, in timer-count units, and communicate that value from the "userland" half of the library to the ISR.
(The library also has `TriacDimmer::detail:ch_A_dn` and `ch_B_dn` that are used in the output compare service routines, to set up the timer to end the trigger pulse at the appropriate time once the pulse has started; this works almost the same but with some further complications around needing to buffer this value to avoid a data race.)

The "userland" end of that communication can be seen in [`TriacDimmer::detail::setChannelA`](https://github.com/AJMansfield/TriacDimmer/blob/master/src/TriacDimmer.cpp#L91-L103) and it's siblings. The code there might not be super readable, due to the `ATOMIC_BLOCK` parts needed to avoid a data race and the logic for ensuring the pulses always end correctly, but if you were to strip out all of that, fundamentally what it's doing is this:

```
void TriacDimmer::detail::setChannelA(float value){
	TriacDimmer::detail::ch_A_up = TriacDimmer::detail::period * value;
}
```

This operation is the key to the library's ability to compensate for different mains and clock frequencies. Because both the period and pulse offset are measured in timer-counts, variation in the absolute size of each timer-count cancel out, and the use of the measured period ensures that the pulse will be proportional to the actual mains frequency.

This is also the reason that there has to be a split between the ISR and "userland" parts of the library. One thing to note, is that the 328p does not have hardware floating point, and in fact it doesn't even have hardware dividers; so an arithmetic operation like this internally involves making a subroutine call into a software floating point routine. This wouldn't _necessarily_ be too expensive to just have the ISR do. But, in general, it's good practice to do as little work as possible in ISRs; and in specific, the library's margins for how close to the start/end of a cycle and how short of a pulse it can generate are directly affected by how long its ISRs take to run, so this arithmetic is done in the "userland" portion.

Because of this, though, this recalibration doesn't happen automatically the way the interrupt-handling and pulse-scheduling parts do; in order to get the library to re-calibrate against any drift in the system clock frequency or mains frequency, these "userland" routines need to be called again, and in theory called regularly in order to _keep_ recalibrating.

In the examples, the main code loops continuously, calling `setBrightness()` regularly (about every 20ms in basic_example.ino, and continuously as it's polling the analog input in potentiometer.ino). Though it doesn't necessarily need to be called continuously like that, it does need to be called at some point _after_ the system has managed to capture and measure a pair of actual zero-crossing events -- i.e. a couple dozen milliseconds after the initial `TriacDimmer::begin()` call.

**Comment by AJMansfield** [comment]
Hmm... tbh writing all of that out I might have an idea for how to fix that limitation of the library, of needing to call into it regularly.

Since, the 328p does have 2-cycle multiply and hardware bitshifts; so if I converted from that user-facing float input into a fixed point binary fraction format, it wouldn't be too expensive to scale the measured period by that value inside the ISR.

No idea if that's actually the issue you're facing without seeing your code, but I'll see if I can publish a new version of the library that can automatically rescale the offsets inside the ISRs.

**Comment by DukesPL** [comment]
I think I found the reason. The problem is in the zero crossing detection system, more precisely in interference from other devices. For example, when I bring the transformer soldering iron close to the optocoupler and click the switch, the light bulb or the motor start to operate unstable. The same situation occurs when I turn on a low-quality switching power supply and bring it closer to the system.
At this point, I placed the PCB in a metal housing, connected to PE (grounding), and this largely limited the problem.
